controllers/swift_controller.go

```diff
@@ -100,53 +100,51 @@ func (r *SwiftReconciler) Reconcile(ctx context.Context, req ctrl.Request) (resu
 		return ctrl.Result{}, err
 	}
 
+	// initialize status if Conditions is nil, but do not reset if it already
+	// exists
+	isNewInstance := instance.Status.Conditions == nil
+	if isNewInstance {
+		instance.Status.Conditions = condition.Conditions{}
+	}
+
+	// Save a copy of the condtions so that we can restore the LastTransitionTime
+	// when a condition's state doesn't change.
+	savedConditions := instance.Status.Conditions.DeepCopy()
+
 	// Always patch the instance status when exiting this function so we can persist any changes.
 	defer func() {
-		// update the Ready condition based on the sub conditions
-		if instance.Status.Conditions.AllSubConditionIsTrue() {
-			instance.Status.Conditions.MarkTrue(
-				condition.ReadyCondition, condition.ReadyMessage)
-		} else {
-			// something is not ready so reset the Ready condition
-			instance.Status.Conditions.MarkUnknown(
-				condition.ReadyCondition, condition.InitReason, condition.ReadyInitMessage)
-			// and recalculate it based on the state of the rest of the conditions
-			instance.Status.Conditions.Set(
-				instance.Status.Conditions.Mirror(condition.ReadyCondition))
-		}
+		instance.Status.Conditions.Set(
+			instance.Status.Conditions.Mirror(condition.ReadyCondition))
```

Comment:
While I don't fully understand the Mirror() function, I do know that it uses LastTransitionTime when determining which condition it mirrors. One thing to try is swapping L116 and L118.

The next thing I'd try is mimicking the other operators, who set ReadyCondiiton to True at the end of a successful reconciliation, rather that trying to determine its value here in the defer function.

Reply:
Ok, I have to think more about this line, and figure out if it's required or not at all: IIUC (@abays correct me if I'm wrong), the defer function is always executed, and even if the `Reconciliation` loop breaks in the middle of its execution for a given reason, this ensures we can "copy" the last failure to the `ReadyCondition`, that is the one that we see when we run for example an `oc get swiftproxy` command.
If we move this to the end, we can only report True or False, but none of the Conditions that might fail in the middle (e.g. Network is not ready).
I'll try to play with it more, and double check if my theory is correct or not, but thank you for checking, I'll try to follow your suggestion too and summarize my findings. 
In general I agree we should try to avoid the pattern of setting conditions in the defer function, but this was an exception for the reason I tried to explain.

Reply:
> the defer function is always executed, and even if the Reconciliation loop breaks in the middle of its execution for a given reason, this ensures we can "copy" the last failure to the ReadyCondition

That is correct :)

Reply:
Hmm, maybe using the Mirror() function can be useful.

Other storage operators recently adopted this pattern:
- Initialize ReadCondition to False with the InitReason at the start of reconciliation
- Set it True at the end of reconciliation if IsReady() is True
- Do not update it in the defer function

So if reconciliation fails or doesn't complete, ReadyCondition will be False but with the InitReason, which isn't very useful.

Alternative approach could be:
- Initialize ReadCondition to Unknown with the InitReason at the start of reconciliation
- Set it True at the end of reconciliation if IsReady() is True
- In the defer function:

```
	defer func() {
		condition.RestoreLastTransitionTimes(&instance.Status.Conditions, savedConditions)
		if instance.Status.Conditions.IsUnknown(condition.ReadyCondition) {
			instance.Status.Conditions.Set(
				instance.Status.Conditions.Mirror(condition.ReadyCondition))
		}
		...
```
This should set ReadyCondition False, but with a reason/severity/message of that last condition that isn't True

Reply:
It might be interesting to have `Unknown` at the beginning of the `Conditions` instead of setting it to `False`. The current implementation starts with `ReadyCondition = False`, and blindly mirrors the Status looking at all the `Conditions` of the `CR`, without checking `isUnknown` or `isFalse`: I assumed it was a redundant logic to have in the `defer` function because of [1] (the underlying `Mirror` function takes care about different use cases).
In particular, in case of failures, [1] sorts the conditions by `LastTransitionTime`, and mirrors the very recent one (that is supposed to be the very last condition that changed over the time) to the `ReadyCondition`.
As per the above, I thought that a blindly call `Mirror` should have been sufficient, but I think I can experiment a bit and follow up. 
Are ok to have a dedicated follow up in a separate PR (built on top of this), so we can move forward with this one that in general takes care about the conditions re-init?

[1] https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/common/condition/funcs.go#L355

Reply:
I have no problem tweaking things in follow-up PRs, but I'm staring at the Mirror() function and see two issues as they pertain to this PR.

1. Because Mirror() relies on sorting conditions in each condition group by their LastTransitionTime [2], it's important to call RestoreLastTransitionTimes() before calling Mirror(). Right?
2. Mirror() uses condition groups, which creates a list of condition lists ordered by importance. False conditions are considered to be most important (and are furthers split into separate SeverityError, SeverityWarning and SeverityInfo groups), then Unknown, and finally True. That's fine, except that I don't see any provision in the code that prevents mirroring a condition with itself. I don't know how this affect things.

[2] https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/common/condition/funcs.go#L352

Reply:
As per the tests in glance-operator [1], I think that the behavior of the `Mirror` function is what we expect to see as long as we initialize `ReadyCondition` to Unknown.
The last patch introduced this behavior and we should be ready to go.


[1] https://github.com/openstack-k8s-operators/glance-operator/pull/496